libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -579,32 +587,48 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			else
 				args.emplace_back(convert(*arguments[i], *parameterTypes[i]).commaSeparatedList());
 
-		if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+		optional<FunctionDefinition const*> functionDef;
+		if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+		{
+			solUnimplementedAssert(!functionType->bound(), "Internal calls to bound functions are not yet implemented for libraries and not allowed for contracts");
+
+			functionDef = dynamic_cast<FunctionDefinition const*>(memberAccess->annotation().referencedDeclaration);
+			solUnimplementedAssert(functionDef.value() != nullptr, "Calls via pointers stored in state variables are not implemented");
+			solAssert(functionDef.value()->isImplemented(), "");
+		}
+		else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
 		{
 			solAssert(!functionType->bound(), "");
-			if (auto functionDef = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
-			{
-				define(_functionCall) <<
-					m_context.enqueueFunctionForCodeGeneration(
-						functionDef->resolveVirtual(m_context.mostDerivedContract())
-					) <<
-					"(" <<
-					joinHumanReadable(args) <<
-					")\n";
-				return;
-			}
+
+			if (auto unresolvedFunctionDef = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
+				functionDef = &unresolvedFunctionDef->resolveVirtual(m_context.mostDerivedContract());
+			else
+				// Not a function name - must be the name of a function pointer variable
+				functionDef = nullptr;
 		}
+		else
+			// Not a simple expression like x or A.x - we must be evaluating a function pointer
```

Comment:
Assumptions like these might be misleading in the future - so I think it's better to just not comment at all.

Reply:
(although the first part of the sentence is valuable)

Reply:
Ok. I'll remove it.

I have to admit that it takes some mental effort to figure out what can and what cannot happen in all those implicit `else`s and `defaults` around the code and I thought it would help to have it documented in some cases. But I know, comments do get out of date pretty easily. Maybe I'll manage to replace it with an assert...

Reply:
As said is the other comment: We cannot harm correctness here, just efficiency. Due to the type of the expression, this is a function that can be called. In the worst case, we go through the internal dispatch, but the right function will be called.

Reply:
Done. The second half of the comment removed. I also added some assertions that verify that we're dealing either with a function definition (static call) or a variable declaration (call via a pointer).